Escaped names compression length mismatch

The following test case fails due to differences in escaping between `PackDomainName` and `Len`:
```go
func TestMsgCompressionEscapingCompressionMap(t *testing.T) {
	msg := new(Msg)
	msg.Compress = true
	msg.SetQuestion("www.example.org.", TypeA)
	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: "ex\\097mple.org.", Rrtype: TypeNS, Class: ClassINET}, Ns: "ns.example.org."})

	predicted := msg.Len()
	buf, err := msg.Pack()
	if err != nil {
		t.Error(err)
	}
	if predicted != len(buf) {
		t.Fatalf("predicted compressed length is wrong: predicted %d, actual %d", predicted, len(buf))
	}
}
```

with:
```
--- FAIL: TestMsgCompressionEscapingCompressionMap (0.00s)
    length_test.go:398: predicted compressed length is wrong: predicted 61, actual 50
```

The problem is that `\\097` is a perfectly valid escaping of `a`, `PackDomainName` handles this by putting the unescaped name (only the first label is unescaped) into the compression map, while the `Len` code puts the escaped name in the compression map. (This isn't effected by #833).

I'm not sure what the best way to fix this is, fixing this in `Len` would (I think) require extra allocations, and fixing this in `PackDomainName` is trickier than it might seem.

This is one of the corner cases I mentioned in https://github.com/miekg/dns/issues/652#issuecomment-442426951.

--- a/dns_bench_test.go
+++ b/dns_bench_test.go
@@ -99,6 +99,15 @@ func BenchmarkMsgLengthOnlyQuestion(b *testing.B) {
 	}
 }
 
+func BenchmarkMsgLengthEscapedName(b *testing.B) {
+	msg := new(Msg)
+	msg.SetQuestion(`\1\2\3\4\5\6\7\8\9\0\1\2\3\4\5\6\7\8\9\0\1\2\3\4\5\6\7\8\9\0\1\2\3\4\5.\1\2\3\4\5\6\7\8.\1\2\3.`, TypeANY)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		msg.Len()
+	}
+}
+
 func BenchmarkPackDomainName(b *testing.B) {
 	name1 := "12345678901234567890123456789012345.12345678.123."
 	buf := make([]byte, len(name1)+1)

--- a/length_test.go
+++ b/length_test.go
@@ -461,7 +461,7 @@ func TestMsgLengthEscaped(t *testing.T) {
 func TestMsgCompressLengthEscaped(t *testing.T) {
 	msg := new(Msg)
 	msg.Compress = true
-	msg.SetQuestion(`www.example.org.`, TypeA)
+	msg.SetQuestion("www.example.org.", TypeA)
 	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: `\000\001\002.example.org.`, Rrtype: TypeNS, Class: ClassINET}, Ns: `ns.\e\x\a\m\p\l\e.org.`})
 	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: `www.\e\x\a\m\p\l\e.org.`, Rrtype: TypeNS, Class: ClassINET}, Ns: "ns.example.org."})
 

--- a/length_test.go
+++ b/length_test.go
@@ -462,7 +462,8 @@ func TestMsgCompressLengthEscaped(t *testing.T) {
 	msg := new(Msg)
 	msg.Compress = true
 	msg.SetQuestion(`www.example.org.`, TypeA)
-	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: `\000\001\002.example.org.`, Rrtype: TypeNS, Class: ClassINET}, Ns: "ns.example.org."})
+	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: `\000\001\002.example.org.`, Rrtype: TypeNS, Class: ClassINET}, Ns: `ns.\e\x\a\m\p\l\e.org.`})
+	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: `www.\e\x\a\m\p\l\e.org.`, Rrtype: TypeNS, Class: ClassINET}, Ns: "ns.example.org."})
 
 	predicted := msg.Len()
 	buf, err := msg.Pack()

--- a/length_test.go
+++ b/length_test.go
@@ -439,9 +439,37 @@ func TestMsgCompressLengthEscapingMatch(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	// Len doesn't account for escaping when calculating the length *yet* so
-	// we're off by three here. This will be fixed in a follow up change.
-	if predicted != len(buf)+3 {
+	if predicted != len(buf) {
+		t.Fatalf("predicted compressed length is wrong: predicted %d, actual %d", predicted, len(buf))
+	}
+}
+
+func TestMsgLengthEscaped(t *testing.T) {
+	msg := new(Msg)
+	msg.SetQuestion(`\000\001\002.\003\004\005\006\007\008\009.\a\b\c.`, TypeA)
+
+	predicted := msg.Len()
+	buf, err := msg.Pack()
+	if err != nil {
+		t.Error(err)
+	}
+	if predicted != len(buf) {
+		t.Fatalf("predicted compressed length is wrong: predicted %d, actual %d", predicted, len(buf))
+	}
+}
+
+func TestMsgCompressLengthEscaped(t *testing.T) {
+	msg := new(Msg)
+	msg.Compress = true
+	msg.SetQuestion(`www.example.org.`, TypeA)
+	msg.Answer = append(msg.Answer, &NS{Hdr: RR_Header{Name: `\000\001\002.example.org.`, Rrtype: TypeNS, Class: ClassINET}, Ns: "ns.example.org."})
+
+	predicted := msg.Len()
+	buf, err := msg.Pack()
+	if err != nil {
+		t.Error(err)
+	}
+	if predicted != len(buf) {
 		t.Fatalf("predicted compressed length is wrong: predicted %d, actual %d", predicted, len(buf))
 	}
 }

--- a/msg.go
+++ b/msg.go
@@ -17,6 +17,7 @@ import (
 	"math/big"
 	"math/rand"
 	"strconv"
+	"strings"
 	"sync"
 )
 
@@ -966,6 +967,10 @@ func domainNameLen(s string, off int, compression map[string]struct{}, compress 
 	}
 
 	nameLen := len(s) + 1
+	if strings.Contains(s, "\\") {
+		nameLen = escapedNameLen(s)
+	}
+
 	if compression == nil {
 		return nameLen
 	}
@@ -974,7 +979,28 @@ func domainNameLen(s string, off int, compression map[string]struct{}, compress 
 		// compressionLenSearch will insert the entry into the compression
 		// map if it doesn't contain it.
 		if l, ok := compressionLenSearch(compression, s, off); ok && compress {
-			nameLen = l + 2
+			// nameLen = l + 2, but this doesn't mess up escapedNameLen
+			// above.
+			nameLen += l - len(s) + 1
+		}
+	}
+
+	return nameLen
+}
+
+func escapedNameLen(s string) int {
+	nameLen := len(s) + 1
+	for i := 0; i < len(s); i++ {
+		if s[i] != '\\' {
+			continue
+		}
+
+		if i+3 < len(s) && isDigit(s[i+1]) && isDigit(s[i+2]) && isDigit(s[i+3]) {
+			nameLen -= 3
+			i += 3
+		} else {
+			nameLen--
+			i++
 		}
 	}
 

--- a/msg.go
+++ b/msg.go
@@ -968,32 +968,27 @@ func domainNameLen(s string, off int, compression map[string]struct{}, compress 
 
 	escaped := strings.Contains(s, "\\")
 
-	nameLen := len(s) + 1
-	if escaped {
-		nameLen = escapedNameLen(s)
-	}
-
-	if compression == nil {
-		return nameLen
-	}
-
-	if compress || off < maxCompressionOffset {
+	if compression != nil && (compress || off < maxCompressionOffset) {
 		// compressionLenSearch will insert the entry into the compression
 		// map if it doesn't contain it.
 		if l, ok := compressionLenSearch(compression, s, off); ok && compress {
 			if escaped {
-				nameLen = escapedNameLen(s[:l]) - 1 + 2
-			} else {
-				nameLen = l + 2
+				return escapedNameLen(s[:l]) + 2
 			}
+
+			return l + 2
 		}
 	}
 
-	return nameLen
+	if escaped {
+		return escapedNameLen(s) + 1
+	}
+
+	return len(s) + 1
 }
 
 func escapedNameLen(s string) int {
-	nameLen := len(s) + 1
+	nameLen := len(s)
 	for i := 0; i < len(s); i++ {
 		if s[i] != '\\' {
 			continue

--- a/msg.go
+++ b/msg.go
@@ -966,8 +966,10 @@ func domainNameLen(s string, off int, compression map[string]struct{}, compress 
 		return 1
 	}
 
+	escaped := strings.Contains(s, "\\")
+
 	nameLen := len(s) + 1
-	if strings.Contains(s, "\\") {
+	if escaped {
 		nameLen = escapedNameLen(s)
 	}
 
@@ -979,9 +981,11 @@ func domainNameLen(s string, off int, compression map[string]struct{}, compress 
 		// compressionLenSearch will insert the entry into the compression
 		// map if it doesn't contain it.
 		if l, ok := compressionLenSearch(compression, s, off); ok && compress {
-			// nameLen = l + 2, but this doesn't mess up escapedNameLen
-			// above.
-			nameLen += l - len(s) + 1
+			if escaped {
+				nameLen = escapedNameLen(s[:l]) - 1 + 2
+			} else {
+				nameLen = l + 2
+			}
 		}
 	}
 

